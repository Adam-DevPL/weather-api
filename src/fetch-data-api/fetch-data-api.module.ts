import { HttpModule } from '@nestjs/axios';
import { Module } from '@nestjs/common';
import { FetchDataApiFilter } from 'src/filters/fetch-data-api.filter';
import { FetchDataApiService } from './fetch-data-api.service';
import { APP_FILTER } from '@nestjs/core';

@Module({
  imports: [HttpModule],
  providers: [
    FetchDataApiService,
    {
      provide: APP_FILTER,
      useClass: FetchDataApiFilter,
    },
  ],
  exports: [FetchDataApiService],
})
export class FetchDataApiModule {}
