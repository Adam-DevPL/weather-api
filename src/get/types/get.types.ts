import { IsAlpha, IsNotEmpty, IsString } from 'class-validator';
import {
  GeoCoordinates,
  LocationType,
} from 'src/prediction/types/prediction.types';

export class GetRouteCountryParam {
  @IsNotEmpty()
  @IsString()
  @IsAlpha()
  country: string;
}

export class GetRouteCityParam {
  @IsNotEmpty()
  @IsString()
  @IsAlpha()
  city: string;
}

export type LocationNameParam = {
  locationType: LocationType.City | LocationType.Country;
  locationName: string;
};

export type CoordinatesParam = {
  lat: number;
  lon: number;
};

export type ApiResponse = {
  avgTemperature: number;
  weather: string;
};

export type WeatherResponse = {
  location: string | GeoCoordinates;
  avgTemperature: number;
  weather: string;
};
