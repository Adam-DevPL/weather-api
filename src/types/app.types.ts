type DateInput = {
  day: string;
};
type DatesRange = {
  from: string;
  to: string;
};

export type GeoCoordinates = {
  latitude: number;
  longitude: number;
};

export type LocationNameInput = {
  locationType: LocationType.City | LocationType.Country;
  locationName: string;
};

export type LocationNameInputDay = LocationNameInput & DateInput;
export type LocationNameInputDates = LocationNameInput & DatesRange;

export type CoordinatesInput = {
  lat: number;
  lon: number;
};

export type CoordinatesInputDates = CoordinatesInput & DatesRange;

export enum LocationType {
  City,
  Country,
  Geo,
}
