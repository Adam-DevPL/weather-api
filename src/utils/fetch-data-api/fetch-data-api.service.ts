import { HttpService } from '@nestjs/axios';
import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { firstValueFrom } from 'rxjs';
import {
  FetchDataApiGeoResponse,
  FetchDataApiParams,
} from './types/FetchDataApi.types';

@Injectable()
export class FetchDataApiService {
  private weatherUrl = 'https://api.open-meteo.com/v1/forecast';
  private geoUrl = 'https://geocoding-api.open-meteo.com/v1/search';

  constructor(private readonly httpService: HttpService) {}

  public async getDataFromApi(params: FetchDataApiParams) {
    const { data } = await firstValueFrom(
      this.httpService.get(this.weatherUrl, { params }),
    );

    return {
      data,
    };
  }

  public async getGeoLocation(name: string): Promise<FetchDataApiGeoResponse> {
    const params = {
      name,
      count: 1,
    };

    const { data } = await firstValueFrom(
      this.httpService.get(this.geoUrl, { params }),
    );

    if (!data.hasOwnProperty('results')) {
      throw new HttpException('Location not found', HttpStatus.BAD_REQUEST);
    }

    const {
      results: [{ latitude, longitude, country }],
    } = data;

    return {
      latitude,
      longitude,
      location: country,
    };
  }
}
