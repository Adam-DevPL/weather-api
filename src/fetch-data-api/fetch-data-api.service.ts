import { HttpService } from '@nestjs/axios';
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import { firstValueFrom } from 'rxjs';
import { FetchDataApiException } from 'src/filters/exceptions/fetch-data-api.exception';
import { LocationType } from 'src/weather/types/weather.types';
import {
  FetchDataApiGeoResponse,
  FetchDataApiParams,
} from './types/FetchDataApi.types';

@Injectable()
export class FetchDataApiService {
  private readonly weatherUrl: string;
  private readonly geoUrl: string;

  constructor(
    private readonly httpService: HttpService,
    private readonly configService: ConfigService,
  ) {
    this.weatherUrl = this.configService.get<string>('WEATHER_URL');
    this.geoUrl = this.configService.get<string>('GEO_URL');
  }

  public async getDataFromApi(params: FetchDataApiParams) {
    try {
      const { data } = await firstValueFrom(
        this.httpService.get(this.weatherUrl, { params }),
      );

      if (data.error) {
        throw new FetchDataApiException('Invalid coordinates');
      }

      return {
        data,
      };
    } catch (error) {
      throw new FetchDataApiException(error.message);
    }
  }

  public async getGeoLocation(name: string): Promise<FetchDataApiGeoResponse> {
    try {
      const params = {
        name,
        count: 1,
      };

      const { data } = await firstValueFrom(
        this.httpService.get(this.geoUrl, { params }),
      );

      if (!data.hasOwnProperty('results')) {
        throw new FetchDataApiException('Location not found');
      }

      const {
        results: [{ latitude, longitude, country }],
      } = data;

      return {
        latitude,
        longitude,
        location:
          country.toLowerCase() === name.toLowerCase()
            ? LocationType.COUNTRY
            : LocationType.CITY,
      };
    } catch (error) {
      throw new FetchDataApiException(error.message);
    }
  }
}
