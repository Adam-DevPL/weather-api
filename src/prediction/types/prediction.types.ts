import { GeoCoordinates } from 'src/types/app.types';

export type ForecastResponse = {
  location: string | GeoCoordinates;
  weatherInfo: WeatherInfo | WeatherInfo[];
};

export type WeatherInfo = {
  day: string;
  avgTemperature: string;
  weather: string;
};

export type OpenModelResponse = {
  data: {
    hourly: {
      time: string[];
      temperature_2m: number[];
      weathercode: number[];
    };
  };
};

export type ApiResponsePredictRoute = {
  timeArr: string[];
  temperatureArr: number[];
  weatherCodeArr: number[];
};
