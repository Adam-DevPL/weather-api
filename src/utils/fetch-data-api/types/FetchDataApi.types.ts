import { LocationType } from 'src/prediction/types/prediction.types';

export type FetchDataApiParams = {
  latitude: number;
  longitude: number;
  start_date?: string;
  end_date?: string;
  current_weather?: true;
  hourly?: 'temperature_2m,weathercode';
};

export type FetchDataApiGeoResponse = {
  latitude: number;
  longitude: number;
  location: LocationType;
};
