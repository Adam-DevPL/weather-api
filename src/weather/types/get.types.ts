import { GeoCoordinates } from 'src/weather/types/weather.types';

export type ApiResponse = {
  avgTemperature: number;
  weather: string;
};

export type WeatherResponse = {
  location: string | GeoCoordinates;
  avgTemperature: number;
  weather: string;
};
