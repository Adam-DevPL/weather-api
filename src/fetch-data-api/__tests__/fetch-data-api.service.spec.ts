import { HttpService } from '@nestjs/axios';
import { HttpException, HttpStatus } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import { Test, TestingModule } from '@nestjs/testing';
import { of } from 'rxjs';
import { FetchDataApiService } from '../fetch-data-api.service';
import { FetchDataApiParams } from '../types/FetchDataApi.types';

describe('FetchDataApiService', () => {
  let service: FetchDataApiService;
  let httpService: HttpService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        FetchDataApiService,
        ConfigService,
        {
          provide: HttpService,
          useValue: {
            get: jest.fn(),
          },
        },
      ],
    }).compile();

    service = module.get<FetchDataApiService>(FetchDataApiService);
    httpService = module.get<HttpService>(HttpService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('getGeoLocation', () => {
    it('should throw error when invalid location', async () => {
      expect.assertions(1);
      //given
      const locationName = 'jhjdhjsd';

      jest.spyOn(httpService, 'get').mockReturnValueOnce(
        of({
          data: {
            invalid: 'invalid',
          },
          headers: {},
          config: { url: 'mockGeoUrl' },
          status: 200,
          statusText: 'OK',
        }) as any,
      );

      //then
      await expect(
        service.getGeoLocation(locationName),
      ).rejects.toMatchSnapshot();
    });

    it('should throw error when problem with external api', async () => {
      expect.assertions(1);
      //given
      const locationName = 'Poland';

      jest.spyOn(httpService, 'get').mockImplementation(() => {
        throw new HttpException('Invalid URL', HttpStatus.BAD_REQUEST);
      });

      //then
      await expect(
        service.getGeoLocation(locationName),
      ).rejects.toMatchSnapshot();
    });
  });

  describe('getDataFromApi', () => {
    it('should throw error when invalid coordinates', async () => {
      expect.assertions(1);
      //given
      const params: FetchDataApiParams = {
        latitude: 200,
        longitude: 200,
        current_weather: true,
      };

      jest.spyOn(httpService, 'get').mockReturnValueOnce(
        of({
          data: {
            error: 'invalid',
          },
          headers: {},
          config: { url: 'mockGeoUrl' },
          status: 200,
          statusText: 'OK',
        }) as any,
      );

      //then
      await expect(service.getDataFromApi(params)).rejects.toMatchSnapshot();
    });

    it('should throw error when problem with external api', async () => {
      expect.assertions(1);
      //given
      const params: FetchDataApiParams = {
        latitude: 200,
        longitude: 200,
        current_weather: true,
      };

      jest.spyOn(httpService, 'get').mockImplementation(() => {
        throw new HttpException('Invalid URL', HttpStatus.BAD_REQUEST);
      });

      //then
      await expect(service.getDataFromApi(params)).rejects.toMatchSnapshot();
    });
  });
});
