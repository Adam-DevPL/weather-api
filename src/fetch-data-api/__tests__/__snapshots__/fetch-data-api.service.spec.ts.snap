// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchDataApiService getGeoLocation should throw error when invalid location 1`] = `
FetchDataApiException {
  "message": "Location not found",
}
`;

exports[`FetchDataApiService should throw error when problem with external api 1`] = `
FetchDataApiException {
  "message": "Invalid URL",
}
`;
